### PR TITLE
fix(shell): persist shell input; prevent tab from blinking; fix telemetry

### PR DESCRIPTION
**Pull request by gribnoysup**
This patch addresses a few issues with the shell tab plugin:

- Use the new browser-repl props in combination with `useTabState` hook to persist the shell input text between tab mounts. This change is very similar to what we were already doing with the output state
- Reduce the tab flicker by removing all of the re-evaluated async code from render. It was kinda okay when the plugin was "global" and only instantiated once, but side-effects are clearly noticeable when it's a tab. There were two things in the code in partuclar that were affecting this:
  - History storage was used directly in the component and reloading itself on every mount. Following the approaches we use in other plugins, I moved this state to the plugin store and as a drive-by just generally cleaned up the store to align it more with a "normal" redux store
  - We were using a slight performance optimisation that was splitting the shell into its own javascript chunk through dynamic import. Even though the chunk is not really loaded from scratch every time, it's still an async operation of retrieving it, adding another few frames to this cascading flow of async render jobs. It would be possible to use some module scope cache to resolve this, but honestly it's just easier to drop this code completely
 - Fix the logging and telemetry by using dedicated event emitter instead of globalAppRegistry. Telemetry for shell is set up by using mongosh/telemetry package that takes an emitter that would be also passed to the runtime and subscribes to the certain events on it, when we are passing the `globalAppRegistry` there, it means that all runtimes across all tabs are setting up their subsriptions on the same emitter and never cleaning them up. Using scoped emitter solves it as it's only used by the created runtime and will be cleaned up when runtime is destroyed.

Kapture 2024-07-03 at 17 48 26

Opening as a draft because this also needs new mongosh version published and updated in Compass first, but otherwise it's ready for review

### CODE REVIEW

File: packages/compass-shell/src/components/compass-shell/compass-shell.tsx [hunk 1]
```diff
@@ -302,9 +231,12 @@ export class CompassShell extends Component<
   }
 }
 
-export default connect((state: RootState) => ({
-  emitShellPluginOpened: () => {
-    state.runtime.appRegistry?.emit('compass:compass-shell:opened');
```
Comment (gribnoysup) [inline packages/compass-shell/src/components/compass-shell/compass-shell.tsx:307]: This event was not used by anything in the app, so I completely removed this emit from both versions of the plugin

File: packages/compass-shell/src/components/shell-header/shell-header.tsx [hunk 2]
```diff
@@ -157,14 +156,6 @@ export const ShellHeader = ({
   );
 };
 
-ShellHeader.propTypes = {
```
Comment (gribnoysup) [inline packages/compass-shell/src/components/shell-header/shell-header.tsx:160]: No point in using prop-types for typescript components

File: packages/compass-shell/src/plugin.tsx [hunk 3]
```diff
@@ -1,91 +1,89 @@
...
+      thunk.withExtraArgument({
+        ...services,
+        historyStorage: new HistoryStorage(),
+      })
+    )
+  );
+
+  setTimeout(() => {
+    void store.dispatch(loadHistory());
+  });
```
Comment (himanshusinghs) [inline packages/compass-shell/src/plugin.tsx:57]: Just curious, why is this and other timeouts needed?

File: packages/compass-shell/src/plugin.tsx [hunk 4]
```diff
@@ -1,91 +1,89 @@
...
-    dataService,
-    preferences,
+
+  addCleanup(destroyRuntime);
+
+  // Trigger terminate when page unloads to kill the spawned child process
+  window.addEventListener('beforeunload', destroyRuntime);
+
+  addCleanup(() => {
+    window.addEventListener('beforeunload', destroyRuntime);
```
Comment (himanshusinghs) [inline packages/compass-shell/src/plugin.tsx:85]:
```suggestion
    window.removeEventListener('beforeunload', destroyRuntime);
```

**Review by himanshusinghs — APPROVED** [review]
Looking great 🚀

File: packages/compass-shell/src/plugin.tsx (hunk 3 above)
Comment (himanshusinghs) [reply packages/compass-shell/src/plugin.tsx:57]: answered over slack

File: packages/compass-shell/src/plugin.tsx (hunk 4 above)
Comment (gribnoysup) [reply packages/compass-shell/src/plugin.tsx]: Thanks for spotting!

**Review by himanshusinghs — APPROVED** [review]

**Pull request merged**